ember-headless-form/src/components/headless-form.hbs

```diff
@@ -1,5 +1,7 @@
 <form
+  novalidate
```

Comment:
This now always suppresses the native way of rendering validation errors, as in:

<img width="235" alt="image" src="https://user-images.githubusercontent.com/1325249/217014256-7263fa1f-63be-4551-867b-a32103921f42.png">

Instead errors are rendered via the existing yielded `field.errors` using any possible custom markup.

Eventually in a future iteration, we might also allow opting into a most simple native-only mode?

Reply:
>  into a most simple native-only mode?

yeah, I think this is pretty important to support. We want to use _T H E  P L A T F O R M_, as much as possible, and while native validation isn't great (dare I said bad? 🙈 ), we're somewhat "breaking the web" by forcing an opt-out.

_instead_, can we only add `novalidate` conditionally? like, is there some way to know that we want to use custom errors?.
Is this worth the trade-offs? maybe an flag to `@nativeErrors` or something? idk

Reply:
(i'm not 100% on my own opinion here, as I think what you have implemented is _the happy path_)

Reply:
Can it be overridden with the `...attributes`? Like if someone adds `novalidate=false`?

Reply:
With [ember-validity-modifier](https://github.com/sukima/ember-validity-modifier#example-validation-on-form-submit) I provide the user with a `verify-form-validity` modifier which will handle both the `novalidate` and `checkValidity()` versus `reportValidity()` checks.

Reply:
Oh interesting! So adding `{{verify-form-validity}}` automatically adds `novalidate`? https://github.com/sukima/ember-validity-modifier/blob/52471db575ec15b852286e3572a94278e8cfe5c4/tests/integration/modifiers/verify-form-validity-test.js#L39

Reply:
> instead, can we only add novalidate conditionally? 

I think we need to add `novalidate` _always_, as our JS backed form needs to be in control when to render validations.  It seems `{{verify-form-validity}}` is also [doing that unconditionally](https://github.com/sukima/ember-validity-modifier/blob/main/addon/-private/validity.js#L163). But that does not mean we cannot do native validation renderings, see below...

> like, is there some way to know that we want to use custom errors?.

Maybe, we could establish some additional "registration protocol" between the field and its yielded `field.errors`, so when `field.errors` is used, that's a signal that users want custom error rendering. However `field.errors` is per field, and the validation mode is for the whole form. Also we might yield also the raw errors array also in addition to the `field.errors` _component_, in which case that's impossible to see if that is used. At the end I would think more of an additional mode you opt-in, I like the `@reportValidity={{true}}` approach that `{{verify-form-validity}}` offers in a similar way.

This matches what we have to do technically as @sukima pointed out: instead of `checkValidity()` (which we currently use) we would have to call `reportValidity()` to trigger the browser to actually show the errors in its native way. In addition to that though, we also have to account for 👇

> Can it be overridden with the ...attributes? Like if someone adds novalidate=false?

Not really, because that would trigger _both_ native rendering as well as our custom rendering (tested it yesterday before our discussion here):

<img width="329" alt="image" src="https://user-images.githubusercontent.com/1325249/217203877-cf54db17-9db6-4c42-8ae9-ce3104e89260.png">

See the same (German, sorry) validation rendered both in native and custom way

So as said above, we need to call `reportValidity()` and suppress yielding our custom errors, which again calls for a dedicated mode like `@reportValidity={{true}}`.

Things start to get more complicated though in that mode, as the question comes up what to do with the JS-based validation callbacks that you can _also_ use to validate fields (`@validate={{validateFn}}` implemented in #21), show they also render in the native way? Which would mean we would have to "register" those errors with the native constraint API using `setCustomValidity()`, which is kinda the opposite way of what we do here (setting native errors from JS, instead of querying native errors from JS)

Given all of that...

> I have yet to see a situation where a designer or a product owner WANTS that native error style

... and that I agree with that, and the need to get something going that at least serves _our current needs_, I was hoping we don't have to bother with this, or at least not _now_. If the need arises, I think we should be able to provide a non-breaking opt-in mode for that.

Reply:
Btw, thanks @sukima for chiming in here! I actually wanted to ping you anyway, given that I was well aware of your work on `ember-validity-modifier`, even before joining CS and getting assigned to this forms project. 

It actually inspired this work, not the actual APIs so much, but more the concept of not doing things very specific to Ember, but leveraging _THE PLATFORM_ (as Preston calls it 🙂). Speaking of the API, this is doing a more component-focused approach vs. the modifier approach you took. The reason for this is that this addon addresses basically two major concerns: validations and accessible markup by default. So it is not truly headless (no idea what the actual definition for this is) in that it does not render _anything_, but it does render things that it has strong opinions on, like again accessible markup (inputs connected to labels, aria attributes etc.), but it makes no assumptions on top of that (no styling, no classes etc.).

By using the Platform, hopefully it is able to compose well with other solutions. E.g. it would nice to see if we could also incorporate custom validation rules just by applying your `{{validity}}` modifier on a headless-form input. There is [this test](https://github.com/CrowdStrike/ember-headless-form/pull/29/files#diff-e54bb85c25006594b458dfc882cd3ab24ddaae50f33ab73705b8200ee394709dR225-R298) with this scenario in mind, so it could already work (though I am not entirely sure how [this](https://github.com/sukima/ember-validity-modifier/blob/main/addon/-private/validity.js#L130-L131) would play out)

So long story short, please share your thoughts here as much as possible and as your time permits. Happy to also have a more high-bandwidth conversation (Zoom) if you want!

Reply:
I had to do a lot more logic workarounds to manage my validity modifier because the components I was using them on lacked `...attributes` which meant the modifier was no longer attached to a specific form field but to a wrapping div and used event bubbling to capture the input/blur/etc. events. Since the _platform_ is taking on the responsibility to manage the `<form>` itself and the form `<input>` fields within providing a modifier for validations would simply be syntactical sugar IMHO.

On to native validity. It isn't reactive in the same way Ember tracked properties are thus you would need an action to trap the causes for the validity to change. Luckily in order to support custom validity you have to do that anyways. My addon did this by trapping the input/change/blur events. But because of bubbling you can do this do this in a wrapping div/component.

To manage the checks and reporting the `<form>` itself keeps a dynamic tally of field elements and can easily provide an iterator for them.

I'm brainstorming here but this is how I might design a system for this. Internally the form component exposes an internal class that tracks validation messages. Attached to the `<form>` by means of an `{{on "validate" …}}` modifier. This form component also traps the `submit` event and based on a flag with either `checkValidity()` or `reportValidity()` before calling the `@onSubmit` action.

Then each field component takes a `@validity` argument which is an array (or singular) of validation functions to call on the inputs’ input/change/blur events. You could leverage a modifier in this case allowing more flexibility in the positional/named params if desired.

In the end ember-validity-modifier can be used _internally_ for the plumbing or you can implement on its own. Either way I think the original event based management is still worth leveraging. Specifically this communication flow:

![diagram of form validation communication flow](https://camo.githubusercontent.com/24976ecd1514decbf63e4e8e8d5a6009015d5cb89ea5f7cb5a6ce9c13aa73e7a/687474703a2f2f7777772e706c616e74756d6c2e636f6d2f706c616e74756d6c2f7376672f58503746496948303343526c564f676d622d6658564f30596967323262794d52355f4b4763635933785248717165656931505f36315f443952456a65374c42507176305f746f5f76435a6b6c7336664e624b61706c6639425771785877644f566e4e544f3265632d784d6b49392d344d71434563326937366a67426f54537a45527a31483654684a46624a493179544a344f67766b67776c4d702d6879654270352d585f612d6c3377467a6650435544786331784f4b726269456d38696f4f6e4652464545616d6d4c2d624855527267616d6967634371304e7237715a724e33643741686646446a4a454173644e67394c6d594a2d6f326d65306d7977734e646a51762d64392d617470354b7833712d797262775557483175586c4b6c35596b49553653794b504d374673432d544f433365565148544a467a757a5859457a61544636733561676941454b38337356427544764f56654a3178367863784458484c764c56)

In the above figure, the Vanilla™ validity and the Ember™ validity are tracked separately but linked by the side effect of the `validated` (past-tense) event. The `validate` (present-tense) is used internally to trigger the attached validation cycles for each field.

```js
let validateEvent = new CustomEvent('validate');
for (let field of form.elements) field.dispatchEvent(validateEvent);
```

And each input component can set its own errors based on that.

```js
@tracked validationMessages = [];

*validate(value, field) {
  let { validity: validators = [] } = this.args;
  if (!Array.isArray(validators)) validators = [validators];
  // validator functions return either [] for valid or […] for invalid
  for (let validator of validators) yield* validator(value, field);
  // Capture Vanilla™ constraint messages
  if (field.validationMessage !== '') yield field.validationMessage;
}

@action
onValidateHandler(event, updatedValue) {
  let field = event.target;
  this.validationMessages = Array.from(this.validate(updatedValue, field));
  field.setCustomValidity(isEmpty(this.validationMessages) ? '' : this.validationMessages[0]);
  field.dispatchEvent(new CustomEvent('validated'), { detail: this.validationMesages });
}
```

I hope this helps.

Reply:
Was also thinking if you liked the ember-validity-modifier, there isn't any technical reason I know of that you could use it as the validation system for this form project. The concept is detached from the use cases. If you look at the [Demo](https://sukima.github.io/ember-validity-modifier/) all you need is to replace `<form>` with your Ember version of things. Including the custom messaging.